core/stat/base/slidingwindow_metric.go

```diff
@@ -0,0 +1,118 @@
+func NewSlidingWindowMetric(sampleCount, intervalInMs uint32, real *BucketLeapArray) *SlidingWindowMetric {
+	if real == nil || intervalInMs <= 0 || sampleCount <= 0 {
+		panic(fmt.Sprintf("Illegal parameters,intervalInMs=%d,sampleCount=%d,real=%+v.", intervalInMs, sampleCount, real))
+	}
+
+	if intervalInMs%sampleCount != 0 {
+		panic(fmt.Sprintf("Invalid parameters, intervalInMs is %d, sampleCount is %d.", intervalInMs, sampleCount))
+	}
+	winLengthInMs := intervalInMs / sampleCount
+
+	parentIntervalInMs := real.IntervalInMs()
+	parentWindowLengthInMs := real.WindowLengthInMs()
+
+	// winLengthInMs of BucketLeapArray must be divisible by winLengthInMs of SlidingWindowMetric
+	// for example: winLengthInMs of BucketLeapArray is 500ms, and winLengthInMs of SlidingWindowMetric is 2000ms
+	// for example: winLengthInMs of BucketLeapArray is 500ms, and winLengthInMs of SlidingWindowMetric is 500ms
+	if winLengthInMs%parentWindowLengthInMs!=0 {
+		panic(fmt.Sprintf("BucketLeapArray's WindowLengthInMs(%d) is not divisible by SlidingWindowMetric's WindowLengthInMs(%d).", parentWindowLengthInMs, winLengthInMs))
+	}
+
+	if intervalInMs > parentIntervalInMs {
+		// todo if SlidingWindowMetric's intervalInMs is greater than BucketLeapArray.
+		panic(fmt.Sprintf("The interval(%d) of SlidingWindowMetric is greater than parent BucketLeapArray(%d).", intervalInMs, parentIntervalInMs))
+	}
+
+	// 10 * 1000 ms == parent
+	if parentIntervalInMs%intervalInMs != 0 {
+		panic(fmt.Sprintf("SlidingWindowMetric's intervalInMs(%d) is not divisible by real BucketLeapArray's intervalInMs(%d).", intervalInMs, parentIntervalInMs))
+	}
+
+	return &SlidingWindowMetric{
+		windowLengthInMs: winLengthInMs,
+		sampleCount:      sampleCount,
+		intervalInMs:     intervalInMs,
+		real:             real,
+	}
+}
+
+// Get the [start time, end time) of SlidingWindowMetric for now.
+func (m *SlidingWindowMetric) getTimeInterval(now uint64) (start, end uint64) {
+	//
+	timeId := now / uint64(m.windowLengthInMs)
+	idx :=  uint32(timeId) % m.sampleCount
+	bucketStartTime := calculateStartTime(now, m.windowLengthInMs)
+
+	start = bucketStartTime - uint64(idx * m.windowLengthInMs)
+	end = start + uint64(m.intervalInMs)
+	return
+}
+
+func (m *SlidingWindowMetric) count(event base.MetricEvent, values []*windowWrap) int64 {
+	ret := int64(0)
+	for _, ww := range values {
+		counter, ok := ww.value.(*metricBucket)
+		if !ok {
+			logger.Errorf("Fail to assert windowWrap value(%+v) to metricBucket.", ww.value)
+			continue
+		}
+		ret += counter.get(event)
+	}
+	return ret
+}
+
+func (m *SlidingWindowMetric) Total() int64 {
+	return m.Get(base.MetricEventPass) + m.Get(base.MetricEventBlock)
+}
+
+func (m *SlidingWindowMetric) Get(event base.MetricEvent) int64 {
+	now := util.CurrentTimeMillis()
+	start, end := m.getTimeInterval(now)
+	satisfiedBuckets := m.real.ValuesWithConditional(now, func(u uint64) bool {
+		return now>=start && now<end
+	})
```

Comment:
Maybe it should be:

```go
satisfiedBuckets := m.real.ValuesWithConditional(now, func(ws uint64) bool {
	return ws>=start && ws<=end
})
```

Reply:
If check [start, end], the final interval is (really interval)+1.
So, use the [start, end)